### DISCUSSION TITLE
Can I pre-compile a large dictionary created though PhraseMatcher?

### DISCUSSION THREAD

**Discussion by lingvisa** [explosion/spaCy#10514] [Help: Coding & Implementations] [answered]
I have a large dictionary created through PhraseMatcher.  The initialization takes around 30 seconds to load the terms into dictionary.

 ```
def __init__(self, nlp, char_match):

        self.matcher = PhraseMatcher(nlp.vocab, attr = 'LOWER')

       ...
       for label, terms in attribute_word_dict.items():
            self.matcher.add(label, list(self.nlp.tokenizer.pipe(terms)))
```

Can I pre-create a dictionary object and hopefully load it later on without going through the same code like above? I am wondering whether and how I can speed up the initialization process.

**Comment by polm** [comment, marked as answer]
There isn't a feature for pre-compilation, and serialization builds the internal state again from text. You might be able to make things faster using Pickle, but that would have to pull in the Vocab/nlp object so I'm not sure how cleanly it would work (it might be fine).

The way you are adding things is a little weird and could maybe be improved. This might be faster:

```
for label, terms in attribute_word_dict.items():
    for term in self.nlp.tokenizer.pipe(terms):
        self.matcher.add(label, [term])
```

Normally reducing calls to `self.matcher.add` might be faster, but if you have a lot of terms then building the list all at once could be causing ineffcient behavior. 

What tokenizer are you using? How big is your dictionary (labels/terms)?

**Reply by lingvisa** [reply]
I have around 800K entries in my dict and it's char tokenizer for Chinese, so the tokenizer doesn't take much time. I switched to code you suggested, and the time now reduced to around 20 seconds from 30 seconds, which is a nice boost.

**Reply by polm** [reply]
800k is more entries than we normally expect to be used with the Matcher, so note that you might just have to deal with it being kind of slow, though 20s startup doesn't sound too bad.

For the specific case of Chinese, since I assume you don't have to worry about spaces much, you might want to look at traditional character-based tries (or DAFSAs or DAWGs), which are often highly optimized and directly support serialization.

**Reply by lingvisa** [reply]
How about the FlashText source code? Or do you have other character-based tries in mind for Chinese that are highly optimized and support serialization? Besides loading time, the  look up time seems fine in my case. My text length cutoff is 400 characters, primarily dealing with social media text.

**Reply by lingvisa** [reply]
I do char based dictionary matching, meaning that both incoming text  doc and word entry doc are char based. After the match, I used tokenization boundary to discard some matches that are inconsistent with boundaries. Is FlashText good for my case?

**Reply by polm** [reply]
I am not super familiar with FlastText, but my understanding is that it's basically a normal trie implementation in Python that makes use of word boundary characters (mainly spaces). I am not sure it works with languages without spaces, you might ask at their repo.

I have used [darts-clone](https://github.com/rixwew/darts-clone-python) before in Japanese tokenizer implementations, but there are many similar libraries.

**Reply by lingvisa** [reply]
Thanks I will look into it.

**Reply by lingvisa** [reply]
I just thought that using python's ProcessPoolExecutor to concurrently. add patterns to Matcher should also be fine, right? If possible, I don't want to switch to another trie library just for the purpose of dictionary initialization.

**Reply by polm** [reply]
I'm not sure there's any way to use multiprocessing to speed up the loading process, since even if nothing breaks you still have to synchronize on the same data structure. 

I would recommend you try pickling the Matcher, and if that doesn't work or doesn't help then you profile it to see what's slow about the creation. Maybe it would help to serialize the Docs you are feeding to the Matcher in a DocBin and load from that? For the number of Docs deserializing will still take a bit though.

I am a little confused about why a 20s startup with that number of patterns is an issue though. Can you explain your use case, particularly your time constraints, a little more?

**Reply by lingvisa** [reply]
I will try pickling PhraseMatcher, and if that works, that would be most straightforward. We had a huge knowledge base, consisting of a large dictionary and a large set of patterns. In initialization, it takes around 20 ms to load the dictionary, another 20 ms to load the patterns, and another 10-15 ms to load bert models. The whole thing takes around 1 minute to initialize and startup. This won't be an issue at all for deployment, but for daily development, it would be great if the startup time can be reduced by half, since developers have to start and end the program frequently.

**Reply by lingvisa** [reply]
Sorry, not 'ms', but  'seconds'.

**Reply by polm** [reply]
If this is only a problem during development I would suggest you use a smaller list of patterns for development.

**Reply by lingvisa** [reply]
Yes, I do have a 'test':true/false setting for loading full or little data to start up, but a lot of times only  when a full KB is loaded, the real effect can be seen and tested, and evaluated. I will trying serialization first. For myself, I actually have got used to this time cost, but for the convenience of others, I hope  to make it faster.
